features/events/api/src/main/java/org/opennms/netmgt/xml/event/IAlarmData.java

```diff
@@ -0,0 +1,48 @@
+/*******************************************************************************
+ * This file is part of OpenNMS(R).
+ *
+ * Copyright (C) 2020 The OpenNMS Group, Inc.
+ * OpenNMS(R) is Copyright (C) 1999-2020 The OpenNMS Group, Inc.
+ *
+ * OpenNMS(R) is a registered trademark of The OpenNMS Group, Inc.
+ *
+ * OpenNMS(R) is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published
+ * by the Free Software Foundation, either version 3 of the License,
+ * or (at your option) any later version.
+ *
+ * OpenNMS(R) is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with OpenNMS(R).  If not, see:
+ *      http://www.gnu.org/licenses/
+ *
+ * For more information contact:
+ *     OpenNMS(R) Licensing <license@opennms.org>
+ *     http://www.opennms.org/
+ *     http://www.opennms.com/
+ *******************************************************************************/
+
+package org.opennms.netmgt.xml.event;
+
+import java.util.List;
+
+/**
+ * A definition corresponding to POJO '{@link AlarmData}'.
+ */
+public interface IAlarmData {
+    Integer getAlarmType();
+    Boolean getAutoClean();
+    String getClearKey();
+    String getReductionKey();
+    String getX733AlarmType();
+    Integer getX733ProbableCause();
+    Boolean isAutoClean();
+    List<IUpdateField> getUpdateFieldList();
+    Boolean hasUpdateFields();
+    IManagedObject getManagedObject();
+    AlarmData toMutableCopy();
```

Comment:
One thing I don't like about this is that it causes the interface to have an explicit dependency on one of the implementations. This would make it difficult if we decided to split these up into separate modules as this module would always have to depend on the implementation.

We could consider if this is worth changing, if it is I don't think there is a convenient solution that avoids changing a bunch of code. One solution would be to remove the copy method from the interface, then add a factory method to the existing mutable bean that takes an interface type and produces a bean copy of that object. That way the decision of typing the mutable object is pushed down to the client, they would explicitly decide they want to get a copy using the JAXB object via its new factory method.

This would apply to all the new interfaces with `toMutableCopy()` methods.

I didn't really think of this possibility when we were originally discussing how to produce these copies.

Reply:
Mapping from mutable to immutable types is now handled in utility class 'Mapper'.

The mapping from immutable to mutable in the 'Mapper' utility class as well... but I'll move these factory methods to the beans themselves (as per your comment).  

The listeners shouldn't need to have knowledge of the 'Mapper'.